### PR TITLE
[Feature] Support pk index major compaction in cloud native table

### LINKED ISSUE #45740: Support pk index major compaction in cloud native table
- [x] https://github.com/StarRocks/starrocks/pull/47112
- [x] https://github.com/StarRocks/starrocks/pull/46002
- [x] https://github.com/StarRocks/starrocks/pull/41737

## Feature request

**Is your feature request related to a problem? Please describe.**

Local table has supported pk index major compaction which will decrease publish timeout. But cloud native table doesn't not support yet.
**Describe the solution you'd like**

Support pk index major compaction in cloud native table.
**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by TszKitLo40**
## Why I'm doing:
PK index major compaction is not supported in cloud native table which will cause write amplification.
## What I'm doing:
Support pk index major compaction in cloud native table.
WIthout pk index major compaction
| | publish cost(ms) | MAX IO(MB/s) |
| :----: | :----: |:----: |
| 1G + insert + 1/100 | 121442 |54 |
| 1G + upsert + 1/100 | 291326 |73 |
| 1G + upsert + 1/10000 | 338894 |46.7 |
| 2.5G + insert + 1/100 | 710491 |121 |
| 2.5G + upsert + 1/100 | 1419680|121 |
| 2.5G + upsert + 1/10000 | 307463|121 |

With pk index major compaction
| | publish cost(ms) | MAX IO(MB/s) |
| :----: | :----: |:----: |
| 1G + insert + 1/100 | 52866 |64.1|
| 1G + upsert + 1/100 | 125956 |80.8 |
| 1G + upsert + 1/10000 | 405559  |46.7 |
| 2.5G + insert + 1/100 | 191436 |117|
| 2.5G + upsert + 1/100 | 389931|108 |
| 2.5G + upsert + 1/10000 | 223999|114 |

In trace log, we can see that the time of `primary_index_commit_latency_us` is much less that without major compaction. Because merge compaction is not needed in commit.
```
I0301 13:51:28.410473  3076 lake_service.cpp:237] Published txns=5050. tablets=10163 cost=1869193us, trace: {"child_traces":[["PublishTablet",{"base_version":93,"deletes":0,"do_update_latency_us":924748,"new_del":0,"primary_index_commit_latency_us":909008,"primary_index_load_latency_us":0,"queuing_latency_us":20,"rewrite_segment_latency_us":19,"rowsetid":93,"state_bytes":9205820,"tablet_id":10163,"total_del":0,"update_index_latency_us":924788,"upsert_rows":151201,"upserts":1}]]}
```

```
I0301 13:52:56.940850  4938 lake_service.cpp:237] Published txns=5031. tablets=10139 cost=11237925us, trace: {"child_traces":[["PublishTablet",{"base_version":102,"deletes":0,"do_update_latency_us":870584,"new_del":0,"primary_index_commit_latency_us":10342835,"primary_index_load_latency_us":0,"queuing_latency_us":16,"rewrite_segment_latency_us":14,"rowsetid":102,"state_bytes":10569270,"tablet_id":10139,"total_del":0,"update_index_latency_us":870614,"upsert_rows":151124,"upserts":1}]]}
```

Fix #45740

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/util/dynamic_cache.h [hunk 1]
```diff
@@ -329,6 +329,20 @@ class DynamicCache {
...
     }
 
+    std::vector<std::pair<Key, Entry*>> get_key_entries() const {
+        std::lock_guard<std::mutex> lg(_lock);
+        std::vector<std::pair<Key, Entry*>> ret;
+        ret.reserve(_map.size());
+        auto itr = _list.begin();
+        while (itr != _list.end()) {
+            Entry* entry = (*itr);
+            entry->_ref++;
```
Comment (luohaha) [inline be/src/util/dynamic_cache.h:339]: When release the ref?

Comment (TszKitLo40) [reply be/src/util/dynamic_cache.h]: Have used a `EntryGuard` to release it.

File: be/src/util/dynamic_cache.h [hunk 2]
```diff
@@ -66,6 +66,18 @@ class DynamicCache {
         T _value;
     };
 
+    class EntryGuard {
+    public:
+        EntryGuard(Entry* entry) : _entry(entry) { _entry->_ref++; }
```
Comment (luohaha) [inline be/src/util/dynamic_cache.h:71]: _ref++ and _ref-- should be protected by _lock

File: be/src/storage/lake/lake_primary_index.h [hunk 3]
```diff
@@ -47,21 +47,27 @@ class LakePrimaryIndex : public PrimaryIndex {
...
 
+    std::timed_mutex* get_index_lock() { return &_mutex; }
+
     void set_enable_persistent_index(bool enable_persistent_index) {
         _enable_persistent_index = enable_persistent_index;
     }
 
+    double get_pk_index_write_amp_score() const { return _pk_index_write_amp_score.load(); }
+
+    void set_pk_index_write_amp_score(double score) { _pk_index_write_amp_score.store(score); }
```
Comment (luohaha) [inline be/src/storage/lake/lake_primary_index.h:69]: execute `set_pk_index_write_amp_score ` after index load

File: be/src/util/dynamic_cache.h [hunk 4]
```diff
@@ -330,6 +330,19 @@ class DynamicCache {
         return _evict(target_capacity, entry_list);
     }
 
+    std::vector<Key> get_keys() const {
```
Comment (luohaha) [inline be/src/util/dynamic_cache.h:333]: Avoid memory copy here.

File: be/src/storage/update_manager.h [hunk 5]
```diff
@@ -38,6 +38,7 @@ class RowsetUpdateState;
 class RowsetColumnUpdateState;
 class Tablet;
 class PersistentIndexCompactionManager;
+using TabletAndScore = std::pair<int64_t, double>;
```
Comment (luohaha) [inline be/src/storage/update_manager.h:41]: Don't define `TabletAndScore` twice, you can define it in `persistent_index_compaction_manager.h`

File: be/src/storage/lake/local_pk_index_manager.cpp [hunk 6]
```diff
@@ -194,6 +200,76 @@ void LocalPkIndexManager::evict(UpdateManager* update_manager, DataDir* data_dir
...
+        UpdateManager* update_manager) {
+    auto tablet_ids = update_manager->index_cache().get_keys();
+    std::vector<TabletAndScore> pick_tablets;
+    if (tablet_ids.empty()) {
+        return pick_tablets;
+    }
+    // 1. pick valid tablet, which score is larger than 0
+    for (auto& tablet_id : tablet_ids) {
+        auto index_entry = update_manager->index_cache().get(tablet_id);
+        DeferOp index_defer([&]() { update_manager->index_cache().release(index_entry); });
```
Comment (luohaha) [inline be/src/storage/lake/local_pk_index_manager.cpp:248]: if (index_entry == nullptr) {
            continue;
        }
        
        DeferOp index_defer([&]() { update_manager->index_cache().release(index_entry); });

**Review by luohaha — DISMISSED** [review]

File: be/src/storage/lake/local_pk_index_manager.cpp [hunk 7]
```diff
@@ -194,6 +200,76 @@ void LocalPkIndexManager::evict(UpdateManager* update_manager, DataDir* data_dir
...
+    for (auto it = _ready_tablets_queue.begin(); it != _ready_tablets_queue.end();) {
+        auto& tablet_score = *it;
+        auto tablet_id = tablet_score.first;
+        if (is_running(tablet_id)) {
+            // remove this tablet because it is already running
+            it = _ready_tablets_queue.erase(it);
+            continue;
+        }
+        auto* data_dir = StorageEngine::instance()->get_persistent_index_store(tablet_id);
+        if (data_dir == nullptr) {
```
Comment (luohaha) [inline be/src/storage/lake/local_pk_index_manager.cpp:214]: it = _ready_tablets_queue.erase(it);

File: be/src/storage/lake/lake_primary_index.h [hunk 8]
```diff
@@ -70,8 +76,9 @@ class LakePrimaryIndex : public PrimaryIndex {
     // We don't support multi version in PrimaryIndex yet, but we will record latest data version for some checking
     int64_t _data_version = 0;
     // make sure at most 1 thread is read or write primary index
-    std::mutex _mutex;
+    std::timed_mutex _mutex;
     bool _enable_persistent_index = false;
+    std::atomic<double> _pk_index_write_amp_score{0.0};
```
Comment (luohaha) [inline be/src/storage/lake/lake_primary_index.h:81]: Removed?

**Review by luohaha — DISMISSED** [review]

**Review by luohaha — DISMISSED** [review]

File: be/src/storage/lake/lake_local_persistent_index.cpp [hunk 9]
```diff
@@ -34,7 +34,7 @@ Status LakeLocalPersistentIndex::load_from_lake_tablet(TabletManager* tablet_mgr
     }
 
     std::unique_ptr<TabletLoader> loader =
-            std::make_unique<LakeLocalPersistentIndexTabletLoader>(tablet_mgr, metadata, base_version, builder);
+            std::make_unique<LakeLocalPersistentIndexTabletLoader>(tablet_mgr, metadata, base_version, builder, this);
```
Comment (kevincai) [inline be/src/storage/lake/lake_local_persistent_index.cpp:37]: is there any way to ensure that the LakeLocalPersisentIndex lives longer than the TabletLoader, so the `this` parameter is always valid during the lifecycle of the loader?

**Review by kevincai — COMMENTED** [review]
to be finished.

File: be/src/storage/lake/update_manager.cpp [hunk 10]
```diff
@@ -353,7 +353,7 @@ Status UpdateManager::_handle_index_op(Tablet* tablet, int64_t base_version, boo
     // release index entry but keep it in cache
     DeferOp release_index_entry([&] { _index_cache.release(index_entry); });
     auto& index = index_entry->value();
-    std::unique_ptr<std::lock_guard<std::mutex>> guard = nullptr;
+    std::unique_ptr<std::lock_guard<std::timed_mutex>> guard = nullptr;
```
[the remark this answers is not in the document]
Comment (TszKitLo40) [reply be/src/storage/lake/update_manager.cpp:356]: We want to reuse some code of local table. And `timed_mutex` is used in the interface of local table.

File: be/src/storage/olap_server.cpp [hunk 11]
```diff
@@ -411,6 +411,12 @@ void* StorageEngine::_pk_index_major_compaction_thread_callback(void* arg) {
             _update_manager->get_pindex_compaction_mgr()->schedule([&]() {
                 return StorageEngine::instance()->tablet_manager()->pick_tablets_to_do_pk_index_major_compaction();
             });
+#ifdef USE_STAROS
+            auto update_manager = ExecEnv::GetInstance()->lake_update_manager();
```
[the remark this answers is not in the document]
Comment (TszKitLo40) [reply be/src/storage/olap_server.cpp:415]: Yes. if major compaction can't catch up. Minor compaction will also stop.

File: be/src/storage/lake/local_pk_index_manager.h [hunk 12]
```diff
@@ -29,15 +30,23 @@ namespace lake {
 class UpdateManager;
 
 #ifdef USE_STAROS
-class LocalPkIndexManager {
+class LocalPkIndexManager : public PersistentIndexCompactionManager {
```
[the remark this answers is not in the document]
Comment (TszKitLo40) [reply be/src/storage/lake/local_pk_index_manager.h:33]: `LocalPkIndexManager` manages the `evict` and `gc` of local persistent index now. And we don't want to add another manager like `LakePersistentIndexCompactionManager` to schedule major compaction. So `LocalPkIndexManager` is reused to do that.

File: be/src/storage/lake/local_pk_index_manager.cpp [hunk 13]
```diff
@@ -194,6 +200,78 @@ void LocalPkIndexManager::evict(UpdateManager* update_manager, DataDir* data_dir
...
+            LOG(ERROR) << strings::Substitute("submit pk index compaction task failed: $0", st.to_string());
+            break;
+        }
+        it = _ready_tablets_queue.erase(it);
+    }
+}
+
+std::vector<TabletAndScore> LocalPkIndexManager::pick_tablets_to_do_pk_index_major_compaction(
+        UpdateManager* update_manager) {
+    auto tablet_ids = update_manager->index_cache().get_keys();
```
[the remark this answers is not in the document]
Comment (TszKitLo40) [reply be/src/storage/lake/local_pk_index_manager.cpp:242]: It is the same as the number of pk tablet which is loading now.

File: be/src/storage/lake/lake_local_persistent_index.cpp (hunk 9 above)
Comment (TszKitLo40) [reply be/src/storage/lake/lake_local_persistent_index.cpp:37]: `loader` is constructed in `load_from_lake_tablet`. And `load_from_lake_tablet` is called by `LakeLocalPersistentIndex` when is loaded. `Loader` will be destructed in `load_from_lake_tablet`.

File: be/src/storage/lake/lake_primary_index.cpp [hunk 14]
```diff
@@ -259,4 +259,25 @@ Status LakePrimaryIndex::major_compact(const TabletMetadata& metadata, TxnLogPB*
     return Status::OK();
 }
 
+double LakePrimaryIndex::get_local_pk_index_write_amp_score() {
+    if (!_enable_persistent_index) {
```
Comment (luohaha) [inline be/src/storage/lake/lake_primary_index.cpp:263]: Use `_persistent_index_type` to replace `_enable_persistent_index`.

**Review by luohaha — APPROVED** [review]

**Review by sevev — APPROVED** [review]

**Review by wyb — APPROVED** [review]

**Comment by TszKitLo40** [comment]
https://github.com/Mergifyio backport branch-3.3

**Pull request merged**